addition of assertSuccess() and assertFailure() on EventLoopFuture

dkz2 opened the pull request:
Added `assertSuccess`, `assertFailure`, `preconditionSuccess` and `preconditionFailure` methods to `EventLoopFuture` for safer and more explicit error handling.

### Motivation:

Enhance developer experience by providing a clear way to handle expected success or failure of futures.

Closes #2359.

### Modifications:

Implemented `assertSuccess`, `assertFailure`, `preconditionSuccess` and `preconditionFailure`  for `EventLoopFuture`.
Added corresponding test cases to verify behavior of the new methods.


### Result:

Developers will now have access to `assertSuccess`, `assertFailure`, `preconditionSuccess` and `preconditionFailure` methods on`EventLoopFuture`.

Lukasa reviewed (changes requested):
Thanks for this! I am inclined to want to say that these methods do slightly too much. Having the callback as an argument muddys the water about their actual semantic. If these methods instead behave like a regular future function (where they return a new future) then we can allow them to compose with the rest of the future handling functions, which might be nicer.

@weissi I'm also curious whether you think assertion is sufficient, or whether we need precondition variants too.

dnadoba commented on Sources/NIOCore/EventLoopFuture.swift:
```diff
@@ -1098,6 +1098,80 @@ extension EventLoopFuture {
+    /// - parameters:
+    ///     - callback: The callback that is called with the successful result of the `EventLoopFuture`.
+    @inlinable
+    @preconcurrency
```
As these are new methods they don't need `@preconcurrency` and also no special treatment in Swift 5.6. We can therefore just have a single declaration and implementation. 

I'm with @Lukasa though and we should make these methods return an `EventLoopFuture` instead. Therefore we will drop the `@Sendable` function argument all together and AFAICT don't have to deal with `Sendable` at all.

dkz2 replied:
Okay i see, that is more for the older methods that were implemented before some language features were introduced,  I will work on addressing the needed changes. Thank you all.

weissi commented:
> Thanks for this! I am inclined to want to say that these methods do slightly too much. Having the callback as an argument muddys the water about their actual semantic. If these methods instead behave like a regular future function (where they return a new future) then we can allow them to compose with the rest of the future handling functions, which might be nicer.

Agreed!

 
> @weissi I'm also curious whether you think assertion is sufficient, or whether we need precondition variants too.

I'd think we should have both `precondition*` and `assert*` variants

dkz2 commented:
> > Thanks for this! I am inclined to want to say that these methods do slightly too much. Having the callback as an argument muddys the water about their actual semantic. If these methods instead behave like a regular future function (where they return a new future) then we can allow them to compose with the rest of the future handling functions, which might be nicer.
> 
> Agreed!
> 
> > @weissi I'm also curious whether you think assertion is sufficient, or whether we need precondition variants too.
> 
> I'd think we should have both `precondition*` and `assert*` variants

Added `precondition*` variants, also removed all unneeded requirements.

Lukasa commented on Sources/NIOCore/EventLoopFuture.swift:
```diff
@@ -2187,6 +2187,79 @@ extension EventLoopFuture {
+            case .success(let value):
+                promise.succeed(value)
+            case .failure(let error):
+                assertionFailure("Expected success, but got failure: \(error)")
```
Can we pass `file` and `line` through from the line that calls `assertSuccess`? That'll make the crash more generally useful.

Lukasa commented on Sources/NIOCore/EventLoopFuture.swift:
```diff
@@ -2187,6 +2187,79 @@ extension EventLoopFuture {
+        self.whenComplete { result in
+            switch result {
+            case .success(let value):
+                assertionFailure("Expected failure, but got success: \(value)")
```
Same note here on `file` and `line`.

glbrntt commented on Sources/NIOCore/EventLoopFuture.swift:
```diff
@@ -2187,6 +2187,79 @@ extension EventLoopFuture {
+    /// If the original future fails, it triggers an assertion failure, causing a runtime error during development.
+    @inlinable
+    public func assertSuccess() -> EventLoopFuture<Value> {
+        let promise = self.eventLoop.makePromise(of: Value.self)
```
No need to create a promise in any of these functions; you can use `always` instead:

```swift
return self.always { result in 
    switch result { 
    case .success:
        ()
    case .failure(let error):
        assertionFailure("Expected success, but got failure: \(error)")    
    }
}
```

weissi replied:
Indeed, but we should pass `file` and `line` to `assertionFailure` such that it shows where the user put `.assertX` and not always this line

dkz2 replied:
oh okay i see no need to create a new promise / or complete it bc `always` can take care of that for us, working on the changes. Thank you!

dkz2 replied:
changes commited!

glbrntt commented on Sources/NIOCore/EventLoopFuture.swift:
```diff
@@ -2187,6 +2187,79 @@ extension EventLoopFuture {
+    ///
+    /// If the original future fails, it triggers a precondition failure, causing a runtime error during development.
+    @inlinable
+    public func ensureSuccess() -> EventLoopFuture<Value> {
```
You can disambiguate by using `Swift.preconditionFailure(...)` to refer to the global one.

dkz2 replied:
Sounds good making the changes!

weissi commented on Sources/NIOCore/EventLoopFuture.swift:
```diff
@@ -2187,6 +2187,79 @@ extension EventLoopFuture {
+    ///
+    /// If the original future fails, it triggers an assertion failure, causing a runtime error during development.
+    @inlinable
+    public func assertSuccess() -> EventLoopFuture<Value> {
```
here and in the other functions we should take `file: ... =  #file, line = ... #line` and pass that to the (assertion/precondition)Failure functions

dkz2 replied:
Done to all!

dkz2 commented on Sources/NIOCore/EventLoopFuture.swift:
```diff
@@ -2187,6 +2187,79 @@ extension EventLoopFuture {
+            case .success(let value):
+                promise.succeed(value)
+            case .failure(let error):
+                preconditionFailure("Expected success, but got failure: \(error)")
```
yes to all!

Lukasa reviewed (approved):
Nice one, this LGTM.

Lukasa commented:
@swift-server-bot test this please

dkz2 commented:
Thank you all for the help and guidance @weissi @Lukasa @glbrntt @dnadoba, very much appreciated!
